File: test/org/intellij/markdown/MarkdownLexerTest.java
package org.intellij.markdown;

import com.intellij.lexer.EmptyLexer;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.testFramework.LexerTestCase;
import org.intellij.markdown.lexer.MarkdownLexer;

public class MarkdownLexerTest extends LexerTestCase {
    @Override protected Lexer createLexer() {
        return new EmptyLexer();
    }

    @Override protected String getDirPath() {
        return "test/data/lexer";
    }

    public void testSimple() {
        defaultTest();
    }

    public void testUnorderedLists() {
        defaultTest();
    }

    public void testOrderedLists() {
        defaultTest();
    }

    public void testHeaders() {
        defaultTest();
    }

    public void testBlockquotes() {
        defaultTest();
    }

    public void testHorizontalRules() {
        defaultTest();
    }

    public void testCodeBlocks() {
        defaultTest();
    }

    public void testCodeFence() {
        defaultTest();
    }

    public void testHtmlBlocks() {
        defaultTest();
    }

    public void testLinkDefinitions() {
        defaultTest();
    }

    public void testCodeSpan() {
        defaultTest();
    }

    public void testLinks() {
        defaultTest();
    }

    public void testAutolinks() {
        defaultTest();
    }

    private void defaultTest() {
        doFileTest("md");
    }

    @Override protected String printTokens(String text, int start) {
        return printTokens(new MarkdownLexer(text));
    }

    public static String printTokens(MarkdownLexer lexer) {

        String result = "";
        while (true) {
            IElementType tokenType = lexer.getType();
            if (tokenType == null) {
                break;
            }
            String tokenText = getTokenText(lexer);
            String tokenTypeName = tokenType.toString();
            String line = tokenTypeName + " ('" + tokenText + "')\n";
            result += line;
            lexer.advance();
        }
        return result;
    }

    private static String getTokenText(MarkdownLexer lexer) {
        String text = lexer.getOriginalText().subSequence(lexer.getTokenStart(), lexer.getTokenEnd()).toString();
        text = StringUtil.replace(text, "\n", "\\n");
        return text;
    }
}

File: test/org/intellij/markdown/MarkdownLexerTest.kt
package org.intellij.markdown;

import org.intellij.markdown.lexer.MarkdownLexer;
import junit.framework.TestCase
import java.io.File

public class MarkdownLexerTest : TestCase() {
    private fun getDirPath() = File("test/data/lexer").getAbsolutePath()

    public fun testSimple() {
        defaultTest();
    }

    public fun testUnorderedLists() {
        defaultTest();
    }

    public fun testOrderedLists() {
        defaultTest();
    }

    public fun testHeaders() {
        defaultTest();
    }

    public fun testBlockquotes() {
        defaultTest();
    }

    public fun testHorizontalRules() {
        defaultTest();
    }

    public fun testCodeBlocks() {
        defaultTest();
    }

    public fun testCodeFence() {
        defaultTest();
    }

    public fun testHtmlBlocks() {
        defaultTest();
    }

    public fun testLinkDefinitions() {
        defaultTest();
    }

    public fun testCodeSpan() {
        defaultTest();
    }

    public fun testLinks() {
        defaultTest();
    }

    public fun testAutolinks() {
        defaultTest();
    }

    public fun testHtmlEntities() {
        defaultTest()
    }

    private fun defaultTest() {
        doFileTest();
    }

    fun printTokens(text: String): String {
        val lexer = MarkdownLexer(text)

        var result = "";
        while (true) {
            val tokenType = lexer.type;
            if (tokenType == null) {
                break;
            }
            val tokenText = getTokenText(lexer);
            val tokenTypeName = tokenType.toString();
            val line = tokenTypeName + " ('" + tokenText + "')";

            if (result.isNotEmpty()) {
                result += "\n";
            }
            result += line;
            lexer.advance();
        }
        return result;
    }

    private fun getTokenText(lexer: MarkdownLexer) = lexer.originalText
            .subSequence(lexer.tokenStart, lexer.tokenEnd)
            .toString()
            .replace("\n", "\\n")

    fun doFileTest() {
        val filePath = getDirPath() + "/" + testName + ".md"
        val result = printTokens(File(filePath).readText())

        assertSameLinesWithFile(getDirPath() + "/" + testName + ".txt", result);
    }
}
